pkg/cmd/container/exec.go

```diff
@@ -0,0 +1,198 @@
+		Client: client,
+		OnFound: func(ctx context.Context, found containerwalker.Found) error {
+			if found.MatchCount > 1 {
+				return fmt.Errorf("multiple IDs found with provided prefix: %s", found.Req)
+			}
+			return execActionWithContainer(ctx, options, args, found.Container, client)
+		},
+	}
+	req := args[0]
+	n, err := walker.Walk(ctx, req)
+	if err != nil {
+		return err
+	} else if n == 0 {
+		return fmt.Errorf("no such container %s", req)
+	}
+	return nil
+}
+
+func execActionWithContainer(ctx context.Context, options types.ExecCommandOptions, args []string, container containerd.Container, client *containerd.Client) error {
+	pspec, err := generateExecProcessSpec(ctx, options, args, container, client)
+	if err != nil {
+		return err
+	}
+
+	task, err := container.Task(ctx, nil)
+	if err != nil {
+		return err
+	}
+	var (
+		ioCreator cio.Creator
+		in        io.Reader
+		stdinC    = &taskutil.StdinCloser{
+			Stdin: os.Stdin,
+		}
+	)
+
+	if options.Interactive {
+		in = stdinC
+	}
+	cioOpts := []cio.Opt{cio.WithStreams(in, os.Stdout, os.Stderr)}
+	if options.TTY {
+		cioOpts = append(cioOpts, cio.WithTerminal)
+	}
+	ioCreator = cio.NewCreator(cioOpts...)
+
+	execID := "exec-" + idgen.GenerateID()
+	process, err := task.Exec(ctx, execID, pspec, ioCreator)
+	if err != nil {
+		return err
+	}
+	stdinC.Closer = func() {
+		process.CloseIO(ctx, containerd.WithStdinCloser)
+	}
+	// if detach, we should not call this defer
+	if !options.Detach {
+		defer process.Delete(ctx)
+	}
+
+	statusC, err := process.Wait(ctx)
+	if err != nil {
+		return err
+	}
+
+	var con console.Console
+	if options.TTY {
+		con = console.Current()
+		defer con.Reset()
+		if err := con.SetRaw(); err != nil {
+			return err
+		}
+	}
+	if !options.Detach {
+		if options.TTY {
+			if err := tasks.HandleConsoleResize(ctx, process, con); err != nil {
+				logrus.WithError(err).Error("console resize")
+			}
+		} else {
+			sigc := commands.ForwardAllSignals(ctx, process)
+			defer commands.StopCatch(sigc)
+		}
+	}
+
+	if err := process.Start(ctx); err != nil {
+		return err
+	}
+	if options.Detach {
+		return nil
+	}
+	status := <-statusC
+	code, _, err := status.Result()
+	if err != nil {
+		return err
+	}
+	if code != 0 {
+		return fmt.Errorf("exec failed with exit code %d", code)
+	}
+	return nil
+}
+
+func withResetAdditionalGIDs() oci.SpecOpts {
+	return func(_ context.Context, _ oci.Client, _ *containers.Container, s *oci.Spec) error {
+		s.Process.User.AdditionalGids = nil
+		return nil
+	}
+}
+
+func generateExecProcessSpec(ctx context.Context, options types.ExecCommandOptions, args []string, container containerd.Container, client *containerd.Client) (*specs.Process, error) {
+	spec, err := container.Spec(ctx)
+	if err != nil {
+		return nil, err
+	}
+	var userOpts []oci.SpecOpts
+	if options.User != "" {
+		userOpts = append(userOpts, oci.WithUser(options.User), withResetAdditionalGIDs(), oci.WithAdditionalGIDs(options.User))
+	}
```

Comment:
For `generateUserOpts`, I think you can refactor [`run_user.go`](https://github.com/containerd/nerdctl/blob/a038f2d532352263fd8438c89131ba0f85aa3f6e/cmd/nerdctl/run_user.go) a little and move the file to `pkg/cmd/container/run_user.go`, so for `generateUserOpts` it will be from

```golang
func generateUserOpts(cmd *cobra.Command) ([]oci.SpecOpts, error)
```

to

```golang
func GenerateUserOpts(user string) ([]oci.SpecOpts, error)
```

Reply:
Then you can use `GenerateUserOpts` in both `exec` (being refactored) as `GenerateUserOpts(opt.User)`, and `run` (not refactored yet) as `container.GenerateUserOpts(user)` where `user` is read from its `cmd`.